### PR TITLE
Feat/secure api

### LINKED ISSUE #9: Secure API

### PR DESCRIPTION
**Pull request by Rexy-Gamaliel**
## Type of Changes

- [ ] Bug fix
- [ ] New feature
- [ ] Code style update
- [x] Refactor
- [ ] Build-related changes
- [x] Add test cases

## Breaking Changes

Is there any breaking changes introduced by this PR?

- [ ] Yes
- [ ] No

If yes, please describe below:

## Description

Resolves #9 by refactoring authentication code flow, now using Authorization in request headers. Also added test cases for all routes

## Checklist

- [x] I have added new tests to cover my changes
- [ ] All new and existing tests are passing

### CODE REVIEW

**Review by stefanusgusega — CHANGES_REQUESTED** [review]
dibenerin dulu ya @Rexy-Gamaliel

File: src/middlewares/auth.middleware.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+export const apiKeyNeeded = (_req: Request, res: Response, next: NextFunction): Response | void => {
+  // const currentApiKey = process.env.API_KEY;
+  // const initialKey = process.env.API_PASSWORD;
+
+  if (!_req.get('API_KEY')) {
+    return res.status(401).send('Unauthorized');
+  }
+  console.log('API_KEY header exist');
+  return next();
+};
```
[the remark this answers is not in the document]
Comment (Rexy-Gamaliel) [reply src/middlewares/auth.middleware.ts:28]: aku coba gitu di `yarn run test` sama lewat Postman ga bisa, bisanya req.get('...')
Di [docs](https://expressjs.com/en/api.html#req.get)-nya juga gaada req.headers

File: src/middlewares/auth.middleware.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+const isValidKey = (currentApiKey: string): boolean => {
+  // const passwordFields = currentApiKey.split('$');
+  // const salt = passwordFields[0];
+  // const hash = createHmac('sha512', salt).update(initialKey).digest('base64');
+  return currentApiKey === 'lorem_ipsum'; // string used as API key
+};
```
[the remark this answers is not in the document]
Comment (Rexy-Gamaliel) [reply src/middlewares/auth.middleware.ts:35]: Barusan dicoba ada kasus menarik bang:

waktu set
`API_KEY="G0/wuYeGu4qBu7SlfIbDfg==$nYJyL8qI80dsXUna/iPOOPMpvS/qF0JuFjohuXag6Ir/XUOYBXyw79klM55QsORp/NZfXEFm5abPSQ/IOzsM3g=="`
di .env.development, pas dibaca nilainya pake `process.env.API_KEY`, substring setelah $ sampai / pertama setelahnya ilang

jadi waktu value `process.env.API_KEY` di auth.middleware.ts nya jadi `G0/wuYeGu4qBu7SlfIbDfg==$iPOOPMpvS/qF0JuFjohuXag6Ir/XUOYBXyw79klM55QsORp/NZfXEFm5abPSQ/IOzsM3g==`

Apa memang itu intensional, atau ada yg salah?

File: src/middlewares/auth.middleware.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (stefanusgusega) [reply src/middlewares/auth.middleware.ts:28]: oke asal bisa jalan gapapa sih rex, karena setelah kulihat harusnya bisa aaja ya pake get

File: src/middlewares/auth.middleware.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  // console.log('Authorization header exist');
+  return next();
+};
+
+const isValidKey = (currentApiKey: string): boolean => {
+  // const passwordFields = currentApiKey.split('$');
+  // const salt = passwordFields[0];
+  // const hash = createHmac('sha512', salt).update(initialKey).digest('base64');
+  // console.log('currentApiKey ' + currentApiKey);
+  // console.log('.env.API_KEY ' + process.env.API_KEY);
```
Comment (stefanusgusega) [inline src/middlewares/auth.middleware.ts:42]: @Rexy-Gamaliel komentar2 yang ga dipake dihapus yaa

File: src/middlewares/auth.middleware.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+  } else {
+    return res.status(401).send('Unauthorized');
+  }
+};
+
+export const apiKeyNeeded = (_req: Request, res: Response, next: NextFunction): Response | void => {
+  // const currentApiKey = process.env.API_KEY;
+  // const initialKey = process.env.API_PASSWORD;
+
+  if (!_req.get('Authorization')) {
```
Comment (stefanusgusega) [inline src/middlewares/auth.middleware.ts:29]: terus _req itu buat variabel req yg ga dipake, jd mending pake req lgsg aja

**Review by stefanusgusega — CHANGES_REQUESTED** [review]

**Pull request merged**